### ISSUE TITLE
FR: Support for `linear="no"` files in EPUB

### ISSUE THREAD

**Issue by Jellby** [koreader/koreader#6809]
The EPUB specification provides for "non-linear" documents, files inside the EPUB that are not included in the normal flow of the text (turning pages from first to last), but that can be accessed though the TOC or hyperlinks. As far as I can see, this property is not supported by KOReader (or most other reading systems), and the non-linear files are treated as any other file. See https://www.w3.org/publishing/epub/epub-packages.html#attrdef-itemref-linear (Note that non-linear documents must still be in the spine, so it is always clear what should be shown when paging forward/backward past their limits.)

**Describe the solution you'd like**
I'd like non-linear documents to be excluded from the normal flow, skipped when paging forward/backward and not counted in the total number of pages for the book or the chapter. If one "visits" these non-linear documents, there should be some indication that the "pages" are now outside the normal flow.

For instance, a book could have several chapters and after the final one a non-linear document with end-notes. When reading the book on page `140/322`, I find a link to an end-note, if I follow the link, I may be in page `[3/18]`. If I page forward to page `[18/18]`, and once again, nothing happens, because that's not the "end of the book" it's just the end of this out-of-flow document. If I page backward to `[1/18]` and one again, I go back to the "normal flow", to page `322/322`. And if I now page forward again, the "end of book" popup appears, because this is the end of the book (the last linear document). Of course I could instead "return to previous position" to go back to page `140/322` and continue reading the main text.

**Describe alternatives you've considered**
Well, the alternative is not doing anything, like currently, and having the book with 322+18 = 340 pages, and after page 322 forcing me to page through 18 pages of end-notes (which I may have already read) to reach the end of the book. Another alternative is excluding the non-linear document from normal paging, but not affecting the number of pages, which would be confusing if the book ends up finishing on page `322/340`.

**Comment by poire-z** [comment]
I have personally no interest in that - and I'd prefer in this case (as in other non-KOReader related contexts) to not be limited to what I'm shown, and see the real thing (all the pages, in their raw order) rather than some kind of view/hide setup a publisher has decided.

> forcing me to page through 18 pages of end-notes (which I may have already read) 

I quite like that, and I do browse thru them, just to see if there is not some easter egg or some other pages than footnotes I have missed (or bug in footnotes and some content that would not have been part of the in-page footnotes, like images, tables...)
And we have the SkimTo widget that helps a lot with that, skipping boring sections, knowing there's stuff after (what would this SkimTo widget show in the case of non-linear stuff?)

Anyway, with crengine/KOReader, we have a single whole linear very-heighty document, and we cut slices in it to make pages. We have a single set of pages (with their top coordinates in the single heighty document), and we access these pages and the top coordinate to show page. That's how we pilot what to show from KOReader. So, this will and should never change.
Any other abstraction should be another map of virtual_pages > real_pages/Xpointers/Y-top-coordinate, a bit like what was done for source page numbers in #6004 (except that in your case, the new page numbers are not absolute but a direct subset, with numbers possibly shifted, of the internal page numbers - and limitations in paging should be done in frontend, and toggable (crengine does not need to know and bother about that).

Some thoughs anyways if one is having a go at that:
- in crengine, we could mark docfragment as linear/hidden - but we should force a page-break when that flag changes between consecutive docfragments, as CSS might not ensure that.
- frontend could then call a function to see if there is non-linear stuff, and if yes, it would fetch the list of pages and their flags (linear/hidden) to build a virtual pagemap
- would need to add hooks in everything that go N page forward/backwards to stop, which would be kinda ugly, for some very rare books.

(I'm sure very little books with footnotes would use that feature to help you from not having to browse thru these footnotes :) so, I wouldn't bother - I bothered with the source page numbers, as this might be to some a kind of valuable additional info.)

> When rendering an EPUB Publication, a Reading System MAY either suppress non-linear content so that it does not appear in the default reading order, or ignore the linear attribute in order to provide users access to the entire content of the EPUB Publication. This specification does not mandate which model Reading Systems have to use. A Reading System MAY also provide the option for users to toggle between the two models.

If one goes at that, I'd replace the last MAY by a SHOULD :)

**Comment by Jellby** [comment]
> I quite like that, and I do browse thru them, just to see if there is not some easter egg or some other pages than footnotes I have missed (or bug in footnotes and some content that would not have been part of the in-page footnotes, like images, tables...)

You're welcome to page through 3000 pages of footnotes in https://www.mobileread.com/forums/showthread.php?p=3972661 ;)

> (I'm sure very little books with footnotes would use that feature to help you from not having to browse thru these footnotes :))

You bet. Since it's an optional feature that is nowhere implemented, as far as I know. But I'm not interested in bought or downloaded books, I'm interested in the books I create or edit, and having a good reading experience on those. And if KOReader could support this, I would enjoy it (I already add `linear="no"` where relevant, although it has no effect).

> > A Reading System MAY also provide the option for users to toggle between the two models.
> 
> If one goes at that, I'd replace the last MAY by a SHOULD :)

I agree that, if implemented, it should be a choice.

**Comment by poire-z** [comment]
Btw, on your 1001 nights book, KOReader complains (but fixes) about:
```
10/21/20-18:08:25 DEBUG validateAndFixToc(): quick scan
10/21/20-18:08:25 DEBUG validateAndFixToc(): TOC needs fixing
10/21/20-18:08:25 DEBUG BOGUS TOC: 241 3774 > 240 4589 - 1 31
10/21/20-18:08:25 DEBUG   fix prev 240 4589 => 3713
10/21/20-18:08:25 INFO  TOC had 1 bogus page numbers: fixed 1 items to keep them ordered.
```
It's one of your `linear=no` page that comes later in the spine vs when it appears in the toc - and we expect/require the TOC to be linearly increasing/monotonic - so we do this dynamic fix to keep it working. (I guess what you did is allowed, but not supported by our TOC code :)
```xml
    <navPoint id="num_239" playOrder="239">
      <navLabel>
        <text>Khelifeh the Fisherman of Baghdad</text>
      </navLabel>
      <content src="night_0831.xhtml#a"/>
    </navPoint>
    <navPoint id="num_240" playOrder="240">
      <navLabel>
        <text>[Variant: Khelif the Fisherman of Baghdad]</text>
      </navLabel>
      <content src="variant.xhtml#a"/>
    </navPoint>
```

**Comment by Jellby** [comment]
Thanks for pointing that out. I never tried it with KOReader, actually, how does it look? I believe I checked whether that was allowed and arrived to the conclusion that it was. But yes, I admit it's a bit clunky. In the paper book the "variant" appears as an appendix to one of the volumes, but with the all-in-one version I didn't want to put it as an appendix to the whole, and I didn't want to force a break in the nights flow.

**Comment by poire-z** [comment]
It looks quite fine.
Some possibly excessive quoting may be :) : 
![image](https://user-images.githubusercontent.com/24273478/96752630-04864b00-13cf-11eb-8cfe-fadcb9f94472.png)

**Comment by Jellby** [comment]
You've seen nothing. Check e.g. Night XXXIII ;) That's what happens with the story-within-story thing.

**Comment by Jellby** [comment]
> * in crengine, we could mark docfragment as linear/hidden - but we should force a page-break when that flag changes between consecutive docfragments, as CSS might not ensure that.

Than part was easy, I could e.g. add a `hidden` attribute to the `<DocFragment>` element and modify the css for extra fun:
```
DocFragment[hidden] {
    page-break-before: always !important;
    page-break-after: always !important;
    background-color: red;
}
```

> * frontend could then call a function to see if there is non-linear stuff, and if yes, it would fetch the list of pages and their flags (linear/hidden) to build a virtual pagemap

That's where I'm getting stuck. How do I know which pages belong to each `<DocFragment>`? Or which `<DocFragment>` contains the first line in a page?

**Comment by poire-z** [comment]
> Some thoughs anyways if one is having a go at that:

That was just some rhetorical thoughts :) I'm not sure I'd want anybody to have a go at that :)
But ok... providing more rhetorical thoughts :/

> How do I know which pages belong to each <DocFragment>? Or which <DocFragment> contains the first line in a page?

You can't currently. And I don't think it should be adressed that way.
There's some bit of code that does page splitting: lvpagesplitter. It's quite complex, tortuous and fragile - and it shouldn't care about xpaths/xpointers and fetch anything from the DOM, so it shouldn't think about accessing the DOM to know about a DocFragment.
After a few minutes thinking, I think the more natural way to do that in the way crengine works would be:

A) EPUB parsing, linear=false => add attribute to DocFragment - what you have already done. No need for the CSS page-break tweaks.

B) Then, in lvrend.cpp/renderBlockEnhanced/FlowState, we walk the DOM. FlowState maintain some state and is a proxy to the page splitting code: renderBlockEnhanced adds "lines/space" (slices of document roll) with properties, and FlowState adds more properties or lines and ensure margin collapsing and page split flags.
I think it should be around there that your cooking should happen:

B1) Have a new member in FlowState to know if we're in a linear=true/false part of the DOM: update/reset it when entering/leaving a DocFragment with linear=false (I would prefer in that code that the state is nonlinear=true, so we can ignore stuff when reading and assume the normal case is "false").

B2) when FlowState is forwarding "lines" to lvpagesplitter/LVRendPageContext, it would forward that flag - and ensure a page break when that state has changed (or may be lvpagesplitter should do that)

B3) lvpagesplitter would have a new member/flag to its "line/lineinfo" objects, and a new member/flag to its "page" objects. When making pages from all the lines it has, (it could force a page split when meeting a new lines where that flag is changed), it would forward that flag to the "page" it is creating and adding.

B4) then, in the list of pages, you have all pages with that new flag: whether the page is made of lines from a non-linear DOM section, or not.

C) After that, I don't know :) we'd need to store somewhere that there are nonlinear pages. Frontend should know that and be able to get the page list (I don't think it ever needed that until now). So, stuff to add to cre.cpp and a bit of glue to lvdocview or lvtinydom to fetch that.

D) After that, it's mostly ugly frontend Lua code :) I have no idea if it can/could/should be handled by the existing ReaderPageMap stuff, or if it would need a new module, or if it should be integrated into core readerrolling (I hope not).
May be that new module could provide a `getInternalPageNumber(cur_page, advance)`, and we would use everywhere we assumed `page_num = page_num + 1`, -1, +10, -10, `page_num = getInternalPageNumber(page_num, -10)` - which would just return `page_num -10` when no linear stuff or when it's disabled, and only do more complex stuff when it's enabled. (Oh my, I don't even want to think about all the edges cases, the footer page number, the TOC items pointing to stuff in non-linear pages, the reading statistics total page accounting, ... :)

**Comment by Jellby** [comment]
> I'm not sure I'd want anybody to have a go at that :)

And I'm not sure I'll get anywhere. I'm not committing myself to implement this or making any promise, I'm just playing around.

> B) Then, in lvrend.cpp/renderBlockEnhanced/FlowState, we walk the DOM. FlowState maintain some state and is a proxy to the page splitting code: renderBlockEnhanced adds "lines/space" (slices of document roll) with properties, and FlowState adds more properties or lines and ensure margin collapsing and page split flags.
> I think it should be around there that your cooking should happen:

Thanks, I think this was the missing piece. I had already guessed/tried to add some property (member) to the lines and pages, but was still missing the connection with the DOM.

> D) After that, it's mostly ugly frontend Lua code :)

I'll worry about that later. I've never done any Lua, so that will be "fun" too.

**Comment by poire-z** [comment]
You might not even need to add any new member (except may be a bool to lvRendPageList, the only thing that stays after rendering and is saved in the cache). Lines and Pages already have flags, and you could just add one bit:
https://github.com/koreader/crengine/blob/5d216f23a9e8e14b27a0417fd014fe021d917515/crengine/include/lvpagesplitter.h#L50-L63
like `RN_LINE_IS_NON_LINEAR 0x2000` and `RN_PAGE_NON_LINEAR 0x40`

(I was going to propose to do B1-B4 and C - as I think it would be non intrusive and cheap - and would need me only a few hours - but you may already be deep in it :) so just say if you need me.)

**Comment by Jellby** [comment]
I think I have something working. Still many things to polish and test, though. I have taken the following approach:
* Page back/forward in "reading mode" skip non-linear fragments
* In skim mode, all pages are accessed, so -10/+10 work as always

I have also added an option to enable/disable the skipping behavior. This should trigger a repagination of the whole document, because of the forced pagebreaks that have to be introduced when enabled. How do I make sure that happens from the lua end? It's similar to changing font / margins / styles, I guess.

Some pending issues:
* I don't think this can be made to work in continuous mode, or is there some way to remove a fragment from the flow and still keep the rest as a continuous "scroll"? (`display: none` would work, but then it cannot be seen at all, not even in footnotes)
* I haven't looked into the 2-page display yet, but I'm hoping this will be easier.

**Comment by poire-z** [comment]
> I don't know how often the linear="no" attribute is used in the wild

On my emulator, among the 366 epubs I have gathered for testing or fixing stuff, 41 have some instance of `linear=.no` in their .opf. So, that makes ~10 % - non negligeable.

>  I think the implementation will be unobtrusive enough once it's polished

Good :)

> you will only see 1 and 3. You can see 2 and 4 by following some link (or maybe they are footnotes), or through the toc (if they are there), or through go to or skim. If you are in fragment 2 and page back, you get first to the beginning of 2, and then to the end of 1. If you page forward instead, you get to the end of and then the beginning of 3. If you are at the end of 1 and page forward, you get to the beginning of 3. If you are in fragment 4 and page forward, you get to the end of 4, and then finish the book. If you are at the end of fragment 3 and page forward, you finish the book.

OK. But once you're in a non-linear fragment (that you can reach only thru something else than page backward/forward), once in there, do page backward and page forward still allow you to page in that non-linear section (so, you can read in there a multi-pages footnote) ?

> But if someone wants to read in paged mode with no extra pagebreaks between chapters, that would be an undesired side effect. 

Note that be default, we have this in epub.css : 
```css
/* EPUB container of each individual HTML file */
DocFragment {
    page-break-before: always;
}
```
The only way to not have this is thru the Style tweaks "Prevent page break at chapter start (or end)".

> I don't think the additional work is that much. It would not be an option you want to toggle very often, at least not more than any other style tweak, I guess, which also need re-rendering.

It's a bit of some ugly work. It would need to be handled as a property of the ldomDocument/tinyNodeCollection, a bit like the global variables that have been made members of the document in https://github.com/koreader/crengine/pull/384 (gHangingPunctuationEnabled is a bool, but does not need a re-render - gInterlineScaleFactor an int, and cause a re-render). Or may be a clearer example https://github.com/koreader/crengine/commit/140044bd2 of how we interface from frontend and provide values (lvdocviewprops, lvdocview, lvtinydom).

Another option, if that's very rare, would be to use an additional -cr-hint CSS named value, so that the additional page-break work can be asked or prevented by just using an additional style tweak:
```css
DocFragment[linear="no" i] {
    -cr-hint: ensure-non-linear-page-breaking;
    /* or if you want it enabled by default : */
    -cr-hint: skip-non-linear-page-breaking;
}
```
That would be a bit less implicit than a proper menu item showing there are non-linear elements and providing options - but that's a solution.

**Comment by Jellby** [comment]
> On my emulator, among the 366 epubs I have gathered for testing or fixing stuff, 41 have some instance of `linear=.no` in their .opf. So, that makes ~10 % - non negligeable.

Then it may also be helpful/useful for others... or it may show that those books are broken :grin: 
 
> OK. But once you're in a non-linear fragment (that you can reach only thru something else than page backward/forward), once in there, do page backward and page forward still allow you to page in that non-linear section (so, you can read in there a multi-pages footnote) ?

Yes, you page normally. Only once you reach the boundary of the fragment you're taken back to the main flow. (So, if you're reading a multi-page footnote, reach the end and page forward without noticing it's the end, you may get the finish book dialog or get carried to some strange place in the book, depending on where exactly this fragment is placed in the spine.)

> The only way to not have this is thru the Style tweaks "Prevent page break at chapter start (or end)".

Right, that was my point. I wouldn't want to "break" those tweaks.

I'll consider your suggestions and see if it's worth it.

**Comment by Jellby** [comment]
Silly question: Is Lua caching my function calls? If so, how can I force it to forget them?

Without going too much into the details, I have something like:
```
function CreDocument:getPageFlow(page)
    print("getPageFlow", page)
    ... some other stuff that calls the crengine interface and returns a value
end
```
This should be called once or several times for each page turn, since the frontend needs to know the flow for each page, so in `readerrolling.lua` there's something like:
```
                test_page = test_page + step
                test_page_flow = self.ui.document:getPageFlow(test_page)
                print("testing page",test_page,"flow",test_page_flow)
```
The point here, I'm calling `getPageFlow` every time before printing `testing page`, so I should get the two prints. But I get this:
```
getPageFlow 3
testing page    3       flow    0
getPageFlow 4
testing page    4       flow    2
getPageFlow 5
testing page    5       flow    0
```
which is good, now if I page back:
```
testing page    4       flow    2
testing page    3       flow    0
```
no `getPageFlow` lines printed.

**Comment by NiLuJe** [comment]
There's a CRe callcache implemented in the Lua frontend, yeah.

I'd say it's in credocument, but I'm not at my dev box right now ;).

In the meantime, you can disable it in the developer options menu ;).

**Comment by poire-z** [comment]
> Silly question: Is Lua caching my function calls? If so, how can I force it to forget them?

Yep, it does:
https://github.com/koreader/koreader/blob/74c1813a8211b721f3e19d9a03977f71c2ee515d/frontend/document/credocument.lua#L1052-L1060

Either disable this (while developping) in File browser > Tools > More tools > [ ] Enable cre call cache.
Or add (now or later, you'll have to later) your new function(s) into the next stuff like:
https://github.com/koreader/koreader/blob/74c1813a8211b721f3e19d9a03977f71c2ee515d/frontend/document/credocument.lua#L1289-L1294
or the other ones depending on how the result can be cached.

**Comment by Jellby** [comment]
Hmm... Thank you both. I'd have to add `getPageCount` too, but that doesn't look clean.
Somehow, when I change the view mode to continuous or paged, the relevant cache entries seem to be cleared/updated, but I can't figure out how/where. If I could do the same when toggling the "hide non-linear" option, that would be preferable, I believe.

**Comment by poire-z** [comment]
> Somehow, when I change the view mode to continuous or paged, the relevant cache entries seem to be cleared/updated, but I can't figure out how/where

That's done by these catch-many-method-names:
https://github.com/koreader/koreader/blob/74c1813a8211b721f3e19d9a03977f71c2ee515d/frontend/document/credocument.lua#L1264-L1270

**Comment by Jellby** [comment]
That was my guess, but I'm calling a `set*` function and I don't get the cache cleared... Anyway, we can maybe discuss it further once I have an implementation to share.
